### PR TITLE
Add a few code optimizations in the TrackDetectorAssociator

### PR DESCRIPTION
**Pull request by perrotta**
#### PR description:

A few rather minor code optimizations are implemented for the TrackDetectorAssociator
In principle this should allow some little speed in computation, but I have to admit that it is barely visible in the performance tests I made

No changes in output are expected. However, there are a few local variables which were copied from a `float` into a local `double`, and then saved again as a `float`, and now they are consistently treated always as a `float`: I don't expect a possible regression also because of them, although it could show up in some remote case.

#### PR validation:

It builds and runs succesfully in a few relevant workflows

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
A test code in the Calibration/HcalCalibAlgos/test area was relying on an include of the TrackDetectorAssociator which was removed by this PR because not needed to stay in the header.
Instead of adding it back to the header of the TrackDetectorAssociator, with the last commit I fixed the remote test code by adding the needed include of CaloGeometryRecord.h

**Comment by perrotta** [comment]
please test

File: TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc [hunk 1]
```diff
@@ -687,28 +653,27 @@ void TrackDetectorAssociator::getTAMuonChamberMatches(std::vector<TAMuonChamberM
       float narrowWidth = wireTopology->narrowWidthOfPlane();
       float length = wireTopology->lengthOfPlane();
       // If slanted, there is no y offset between local origin and symmetry center of wire plane
-      float yOfFirstWire = fabs(wireTopology->wireAngle()) > 1.E-06 ? -0.5 * length : wireTopology->yOfWire(1);
+      float yOfFirstWire = std::abs(wireTopology->wireAngle()) > 1.E-06 ? -0.5 * length : wireTopology->yOfWire(1);
```
Comment (VinInn) [inline TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc:656]: add "f" to the literals (w/o are considered double)

File: TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc [hunk 2]
```diff
@@ -687,28 +653,27 @@ void TrackDetectorAssociator::getTAMuonChamberMatches(std::vector<TAMuonChamberM
...
-      float yPrime = localPoint.y() + fabs(yOfFirstWire);
+      float yPrime = localPoint.y() + std::abs(yOfFirstWire);
       // half trapezoid width at y' is 0.5 * narrowWidth + x side of triangle with the above tangent and side y'
       float halfWidthAtYPrime = 0.5 * narrowWidth + yPrime * tangent;
-      distanceX = fabs(localPoint.x()) - halfWidthAtYPrime;
-      distanceY = fabs(localPoint.y() - yCOWPOffset) - 0.5 * length;
-      sigmaX = distanceX / sqrt(localError.xx());
-      sigmaY = distanceY / sqrt(localError.yy());
+      distanceX = std::abs(localPoint.x()) - halfWidthAtYPrime;
+      distanceY = std::abs(localPoint.y() - yCOWPOffset) - 0.5 * length;
```
Comment (VinInn) [inline TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc:667]: ditto

Comment (VinInn) [reply TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc:667]: also in tangent =

File: TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc [hunk 3]
```diff
@@ -848,17 +814,17 @@ bool TrackDetectorAssociator::addTAMuonSegmentMatch(TAMuonChamberMatch& matchedC
   if (dtseg && (!dtseg->hasZed()))
     isDTWithoutY = true;
 
-  double deltaPhi(fabs(segmentGlobalPosition.phi() - trajectoryStateOnSurface.freeState()->position().phi()));
+  float deltaPhi(std::abs(segmentGlobalPosition.phi() - trajectoryStateOnSurface.freeState()->position().phi()));
   if (deltaPhi > M_PI)
-    deltaPhi = fabs(deltaPhi - M_PI * 2.);
```
Comment (VinInn) [inline TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc:853]: M_PI is double and 2. as well

File: TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc [hunk 4]
```diff
@@ -1051,7 +1017,7 @@ TrackDetMatchInfo TrackDetectorAssociator::associate(const edm::Event& iEvent,
         // check if we deal with clear outside-in case
         if (track.innerPosition().Dot(track.innerMomentum()) < 0 &&
             track.outerPosition().Dot(track.outerMomentum()) < 0) {
-          cachedTrajectory_.setPropagationStep(-fabs(currentStepSize));
+          cachedTrajectory_.setPropagationStep(-std::abs(currentStepSize));
           TrackDetMatchInfo result;
           if (track.innerPosition().R() < track.outerPosition().R())
```
Comment (VinInn) [inline TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc:1022]: R2 (or whatever equivalent)?

**Comment by perrotta** [comment]
Litterals added, and `PositionVector3D::R()` replaced by `PositionVector3D::Mag2()`: thank you @VinInn

**Comment by perrotta** [comment]
please test

File: TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc [hunk 5]
```diff
@@ -979,10 +945,10 @@ void TrackDetectorAssociator::fillCaloTruth(const edm::Event& iEvent,
     info.ecalTrueEnergy = ecalTrueEnergy;
     info.hcalTrueEnergy = hcalTrueEnergy;
     info.hcalTrueEnergyCorrected = hcalTrueEnergy;
-    if (fabs(info.trkGlobPosAtHcal.eta()) < 1.3)
-      info.hcalTrueEnergyCorrected = hcalTrueEnergy * 113.2;
-    else if (fabs(info.trkGlobPosAtHcal.eta()) < 3.0)
-      info.hcalTrueEnergyCorrected = hcalTrueEnergy * 167.2;
+    if (std::abs(info.trkGlobPosAtHcal.eta()) < 1.3f)
+      info.hcalTrueEnergyCorrected = hcalTrueEnergy * 113.2f;
```
Comment (perrotta) [inline TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc:949]: This class contains quite several hardcoded parameters, which are prone to get out of date and hard to maintain. These caliibration factors for the hcal energy are one example (@abdoulline @igv4321 FYI), but there are several more around.
It is not the purpose of this PR, but I think that at some point one should make them either configurable, or taken from some central database or code.

Comment (abdoulline) [reply TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc:949]: Well, these hardcoded HCAL scaling factors can be replaced with config.ones, but it requires some (importing) gymnastics: 

(1) here in  the  TrackDetectorAssociator.cc  from hit = simHcalHits  one can get  
HcalDetId id;   like this 
https://cmssdt.cern.ch/lxr/source/Validation/HcalHits/src/SimHitsValidationHcal.cc#0172
 
(2) 
int ieta = id.ieta();   int subdet = id.subdet();  
where the latter is 1 (HcalGenericDetId::HcalGenBarrel)  or 2 (HcalGenericDetId::HcalGenEndcap) as enumerated in 
https://cmssdt.cern.ch/lxr/source/DataFormats/HcalDetId/interface/HcalGenericDetId.h
 
(3) there are "Sampling factors" separately for HB and HE  (for approx. estimate of SimHits energy scale -> Reco scale used in HCAL Digitization):
https://cmssdt.cern.ch/lxr/source/SimCalorimetry/HcalSimProducers/python/hcalSimParameters_cfi.py#0066
https://cmssdt.cern.ch/lxr/source/SimCalorimetry/HcalSimProducers/python/hcalSimParameters_cfi.py#0085

for |ieta|=1-16  and subdet==HcalGenericDetId::HcalGenBarrel and 
for |ieta|=16-29 and subdet==HcalGenericDetId::HcalGenEndcap
as specified in 
https://cmssdt.cern.ch/lxr/source/SimCalorimetry/HcalSimProducers/python/hcalSimParameters_cfi.py#0062
https://cmssdt.cern.ch/lxr/source/SimCalorimetry/HcalSimProducers/python/hcalSimParameters_cfi.py#0081   
repectively

**Comment by perrotta** [comment]
+1
- Just a few code optimizations, that do not change the outcome of the associator (and which also bring only some very minor improvement to the overall timing of the code, though...)
- Jenkins test pass with no differences, as expected

**Comment by perrotta** [comment]
@cms-sw/alca-l2 the alca part of this PR only  consists in adding a previously missing include in a test script: I think this can be safely signed by alca

**Comment by christopheralanwest** [comment]
+alca

**Comment by silviodonato** [comment]
+1

**Pull request merged**